src/wp-admin/includes/file.php

```diff
@@ -1746,7 +1749,18 @@ function _unzip_file_ziparchive( $file, $to, $needed_dirs = array() ) {
 			return new WP_Error( 'mkdir_failed_ziparchive', __( 'Could not create directory.' ), $_dir );
 		}
 	}
-	unset( $needed_dirs );
+
+	/**
+	 * Fires before a ZIP archive is extracted.
+	 *
+	 * @since 6.3.0
+	 *
+	 * @param string   $file           Full path and filename of ZIP archive.
+	 * @param string   $to             Full path on the filesystem to extract archive to.
+	 * @param string[] $needed_dirs    A full list of required folders that need to be created.
+	 * @param float    $required_space The space required to unzip the file and copy its contents, with a 10% buffer.
+	 */
+	do_action( 'before_unzip_file', $file, $to, $needed_dirs, $required_space );
```

Comment:
After a Slack discussion with @peterwilsoncc, these hooks may be better as filters to cover the use cases mentioned in the ticket such as malware scanning and such.

i.e.
```php
/**
 * Filters whether to override unzipping an archive.
 *
 * @since 6.4.0
 *
 * @param null|true|WP_Error $result         The result of the override. True on success, otherwise WP_Error. Default null. 
 * @param string             $file           Full path and filename of the ZIP archive.
 * @param string             $to             Full path on the filesystem to extract the archive to.
 * @param string[]           $needed_dirs    A full list of required folders that need to be created.
 * @param float              $required_space The space required to unzip the file and copy its contents, with a 10% buffer.
 */
$pre = apply_filters( 'pre_unzip_file', null, $file, $to, $needed_dirs, $required_space );

if ( null !== $pre ) {
    return $pre;
}
```
and
```php
/**
 * Filters the result of unzipping an archive.
 *
 * @since 6.4.0
 *
 * @param true|WP_Error $result         The result of unzipping the archive. True on success, otherwise WP_Error. Default true.
 * @param string        $file           Full path and filename of the ZIP archive.
 * @param string        $to             Full path on the filesystem the archive was extracted to.
 * @param string[]      $needed_dirs    A full list of required folders that were created.
 * @param float         $required_space The space required to unzip the file and copy its contents, with a 10% buffer.
 */
$result = apply_filters( 'unzip_file', true, $file, $to, $needed_dirs, $required_space );

unset( $needed_dirs );

return $result;
```

@azaozz  @SergeyBiryukov What do you think?

Reply:
After further consideration, I think filters meet the needs of the ticket much better than actions. Updated in [6879aab](https://github.com/WordPress/wordpress-develop/pull/4220/commits/6879aabbb491642b913b14d274f9e558e3e574be).